### ISSUE TITLE
Support react-native-document-picker

### ISSUE THREAD

**Issue by irakaundal** [microsoft/react-native-windows#4209]
# Support `react-native-document-picker` on RNW vNext

## Basic Info
- Module Repo URL: https://github.com/Elyx0/react-native-document-picker
- Target version: 2.0.0
- [ ] Issue exists on module repo (if so, [link]())

## Overview of Required Functionality

Need to add c++/winrt support for the npm package so that it can work with the latest version of react-native-windows.

**Comment by stmoy** [comment]
Hi @irakaundal - thank you for filing this request. To help with prioritization, can you provide some information about which app/product you work on?

**Comment by irakaundal** [comment]
The project we needed it for has been halted right now.

**Comment by creambyemute** [comment]
@stmoy @chrisglein I've stumbled across this one this week when evaluating if there's any existing file-picker library that supports RN-Windows. 
We are using react-native-image-picker on iOS/Android and thought we could add rn-document-picker for Windows.

I guess the old implementation isn't linkable with a RN-Windows VNext/C++ App? At least I couldn't get it building/linking correctly.

I guess my best bet is to follow the instructions from @namrog84 from https://github.com/microsoft/react-native-windows/issues/4255 and create a small file/image picker module myself? 

Like described in the issue
```
// Pick file
StorageFile file = co_await filePicker.PickSingleFileAsync();

// Copy to temp directory
newPath = winrt::Windows::Storage::ApplicationData::Current().TemporaryFolder()

// return newPath for later use outside the picker
```
Or are you aware of something ready to use?

**Comment by chrisglein** [comment]
> I guess the old implementation isn't linkable with a RN-Windows VNext/C++ App? At least I couldn't get it building/linking correctly.

Correct, there are legacy implementations that aren't compatible with current versions of React Native for Windows. There are a handful of those still around, generally being converted as people like yourself find a need for them. Given that there's a reference implementation it's very possible to use that code and get it up and running again. @jonthysell has experience with this and can link you to docs / example PRs (@stmoy might also have a list of examples that have been tackled).

Outside of that, yes your next best option is to drop in and add the native code you need to your own project, as you've outlined.

**Comment by namrog84** [comment]
There is another new feature in RNW that was added after my original issue was filed that might help address this and that is the new ReactContext.Properties() bag.
e.g. https://github.com/microsoft/react-native-windows/blob/master/packages/sample-apps/windows/SampleLibraryCPP/SampleModuleCPP.h#L42

Since changing Windows APIs is unlikely.  The next best course of action, would be to agree on some common key dictionary key "names" for a few properties for things like  "LastAccessedFileStorage" and store the StorageFile pointer inside this property bag via the file picker. 
Then in subsequent other modules if "LastAccessedFileStorage"'s path matched the string wanting to be used, it'd be able to retrieve the original StorageFile pointer that has proper access to anywhere. 

The SampleModuleCpp's example only is showing an `int` and a `winrt::hstring` being stored in the PropertiesBag, but I believe it has support for arbitrary void obj pointers or something more flexible than primitives. 

I think @vmoroz implemented it and might be able to better address if this is a good use case for that feature.

**Comment by creambyemute** [comment]
Alright, thanks for the heads up.

I've just successfully created my first C++ Native module which uses `Launcher::LaunchFileAsync` with `LauncherOptions` to open local files.

I'll give the FilePicker a try next.

**Comment by petr7555** [comment]
Hi.
I would like to get the same functionality as react-native-document-picker offers. I have spent two full days trying to get it working but without any luck. The documentation is very unclear about how should I work with native modules.

I have an app, which has `windows` folder. This was generated by `npx react-native-windows-init --overwrite`. It contains a solution made of the **main project** (same name as the solution) and referenced projects (DateTimePickerWindows, RNPermissions etc.).
I was able to add the FilePicker _creambyemute_ posted to my main c++ project as FilePicker.h and FilePicker.cpp and it was picked automatically and I could use it in JS. I guess this is the **auto-linking**? 
I created a new repository with https://github.com/callstack/react-native-builder-bob and wanted to add a windows support with `npx react-native-windows-init --version latest --projectType lib --overwrite`. But this command, despite the lib flag, generates an app. 
So I wanted to create the windows folder manually (following the structure of https://github.com/zoontek/react-native-permissions/tree/master/windows/RNPermissions). This is very hard as I do not know which dependencies I need. What does it take for react native windows to "autolink" modules from such project? Does it need to be added and referenced from the **main project**? Do I also need to add the header file to `pch.h` and `PackageProviders().Append(winrt::FilePicker::ReactPackageProvider());` to `App.cpp`? The documentation says this is done automatically, but then it gives instructions on how to do it manually. Why?! How else would I say where to look for native modules? Are node_modules somehow scanned for REACT_MODULE macros? 
So I abandoned the idea of having a separate File Picker library and decided to put all the functionality directly inside the main project.

I wanted to extend the functionality of FilePicker and make use of JSValueObject options. But I could not find out how to get JSValueArray out of it.
I was not able to find any c++ documentation for FileOpenPicker.
The documentation says c++ project is preferred over c# one, but all documentation I could find was for c# code.

Therefore I would like to add a c# project to my current c++ solution. Write a native module inside that project and have it picked up automatically by my app and use it in JS. But the documentation only tells how to link c++ projects https://microsoft.github.io/react-native-windows/docs/native-modules-using. I do not have any header file in my c# project. Furthermore, this should manual linking should not be necessary, as it is supposed to be done "automatically".

I am sorry for such confusing description, but I am myself VERY confused. The only thing I want to achieve is to be able to get a base64 string of image on PC. However, React Native Windows and its documentation make this task impossible so far.

**Comment by creambyemute** [comment]
@petr7555 The documentation of creating an autolinking external c++ module imo is a bit all over the place. 
I also haven't finished creating a native module out of it yet but there is a pretty lenghty description somewhere which you have to follow carefully. 
And I also faced similiar problems with the documentation of Windows APIS for C++/WinRT. Some had them some not.

**Comment by vmoroz** [comment]
@petr7555, I am sorry to hear that you are having so many issues. I will be glad to help you with some of them.

@jonthysell knows all about the auto-linking implementation. My understanding is that the entry point for auto-linking starts with the package.json dependency and then it generates a file that registers the package.
Each DLL must have an entry point for an app to use its native modules. In our case it is the implementation of the IReactPackageProvider interface. Each REACT_MODULE creates a static entry which can be used inside of the ReactPackageProvider::CreatePackage method implementation. We can use AddAttributedModules method to add to the package all static entries created by REACT_MODULE attributes, or TryAddAttributedModule to add just one module with a specific name. See the sample where we register all attributed modules, a TurboModule, and custom views in the package: [ReactPackageProvider.cpp](https://github.com/microsoft/react-native-windows/blob/master/packages/sample-apps/windows/SampleLibraryCPP/ReactPackageProvider.cpp)

To create a new module I usually use another project as a starting point, change there project GUID and change all the required names and references. I used to use the sample module project: [SampleLibraryCPP](https://github.com/microsoft/react-native-windows/tree/master/packages/sample-apps/windows/SampleLibraryCPP), but it may have some settings specific for RNW source tree. One of the latest C++ DLL I helped to setup recently was [RNFetchBlob](https://github.com/avmoroz/rn-fetch-blob/tree/dev/windows/RNFetchBlob). You may consider to start with this one, but still see what we do in the sample project. RNFetchBlob project was done before RNW 0.64 release and could be outdated.

One of the difficult part of the extension module development is to test it with your app while both of them are still under development. You may either use a sample app approach as Andrei did in [RNFetchBlobWin](https://github.com/avmoroz/rn-fetch-blob/tree/dev/examples/RNFetchBlobWin/windows) and/or to use [Verdaccio](https://verdaccio.org/). I found that [Verdaccio](https://verdaccio.org/) is the only way to verify that the new module works with any app. Since it takes extra steps to redeploy changes, I usually use it towards the completion of the project. Also, please see the docs about the native module development: [native-modules-setup](https://microsoft.github.io/react-native-windows/docs/native-modules-setup).

For C++ vs C# modules. Please do not create C# modules - there may be difficulties to deploy them with C++ apps. See [native-code-language-choice](https://microsoft.github.io/react-native-windows/docs/native-code-language-choice).

For the JSValue/JSValueObject/JSValueArray: the RNW modules use ABI-safe API based on the IJSValueReader and IJSValueWriter interfaces, and in 0.64 the experimental JSI implementation. The JSValue is one of the ways to send data using IJSValueReader and IJSValueWriter. The recommendation is to use specific types and implement ReadValue/WriteValue methods for them. The JSValue is quite similar to folly::dynamic to help implementing ReadValue in complex cases. JSValue is designed to be immutable to enable safe transferring values between threads. The JSValueObject and JSValueArray are used as JSValue builders for Object and Array values, and to access their immutable contents. I still need to write detailed docs about them, but you can see some of the docs here [native-modules-jsvalue](https://microsoft.github.io/react-native-windows/docs/native-modules-jsvalue), and the usage in our unit test projects: [JSValueTest.cpp](https://github.com/microsoft/react-native-windows/blob/master/vnext/Microsoft.ReactNative.Cxx.UnitTests/JSValueTest.cpp). Custom ReadValue/WriteValue implementation: [JSValueReaderTest.cpp](https://github.com/microsoft/react-native-windows/blob/master/vnext/Microsoft.ReactNative.Cxx.UnitTests/JSValueReaderTest.cpp).

**Comment by petr7555** [comment]
@vmoroz Thank you very much for your help. 

I abandoned the idea of having it as an external dependency and stuck with FilePicker.cpp and FilePicker.h directly inside my app. I used parts of code from [RNFetchBlobWin](https://github.com/avmoroz/rn-fetch-blob/tree/dev/examples/RNFetchBlobWin/windows) to implement the "file to base64 functionality". [Here's](https://github.com/petr7555/bachelors_thesis_accounting_ocr/blob/master/windows/bachelors_thesis_accounting_ocr/FilePicker.cpp) the code if anyone needed it.

I think the documentation should be more clear about c++ being preferred, and in the future perhaps the only way to go. Based on the table [here](https://microsoft.github.io/react-native-windows/docs/native-code-language-choice) I understand, that it is not possible to add a c# project to c++ RN Windows project. I guess this is the reason why [react-native-document-picker](https://github.com/rnmods/react-native-document-picker) does not work anymore...?

On a high level and based on my experience my understanding of auto-linking is that for some modules it works and the module is linked during "npm run windows" but sometimes it does not and the module needs to be linked manually (by adding the module as a reference and changing App.cpp and pch.h files).

Maybe someone could look into why `npx react-native-windows-init --version latest --projectType lib` generates app and not lib? Copying from other modules seems cumbersome.
